kms-connector/crates/kms-worker/src/core/event_processor/s3.rs

```diff
@@ -211,70 +219,137 @@ where
                     Ok(ciphertext) => {
                         info!(
-                            "Successfully retrieved ciphertext for digest {} from S3 URL {}",
-                            alloy::hex::encode(&extracted_sns_ciphertext_digest),
-                            s3_url
-                        );
-                        sns_ciphertext_materials.push((extracted_ct_handle.clone(), ciphertext));
-                        ciphertext_retrieved = true;
-                        break; // We want to stop as soon as ciphertext corresponding to extracted_sns_ciphertext_digest is retrieved
-                    }
-                    Err(error) => {
-                        // Log warning but continue trying other URLs
-                        warn!(
-                            "Failed to retrieve ciphertext for digest {} from S3 URL {}: {}",
-                            alloy::hex::encode(&extracted_sns_ciphertext_digest),
-                            s3_url,
-                            error
+                            attempt = i,
+                            "Successfully retrieved ciphertext for digest {digest_hex} from S3 URL {s3_url}"
                         );
-                        // Continue to the next URL
+                        return Some(ciphertext);
                     }
+                    Err(e) => warn!(
+                        attempt = i,
+                        "Failed to retrieve ciphertext for digest {digest_hex} from S3 URL {s3_url}: {e}"
+                    ),
                 }
             }
+        }
+        None
+    }
 
-            if !ciphertext_retrieved {
-                warn!(
-                    "Failed to retrieve ciphertext for digest {} from any S3 URL",
-                    alloy::hex::encode(&extracted_sns_ciphertext_digest)
-                );
-                // Continue to the next SNS material
-            }
+    /// Retrieves a ciphertext from S3 using the bucket URLs and ciphertext digest.
+    pub async fn retrieve_s3_ciphertext(
+        &self,
+        s3_bucket_url: &str,
+        ciphertext_digest: &[u8],
+        digest_hex: &str,
+    ) -> anyhow::Result<Vec<u8>> {
+        info!(
+            "S3 RETRIEVAL START: Retrieving ciphertext with digest {} from S3 bucket {}",
+            digest_hex, s3_bucket_url
+        );
+
+        // Direct HTTP retrieval
+        let direct_url = format!("{s3_bucket_url}/{digest_hex}");
+        let ciphertext = self.direct_http_retrieval(&direct_url).await?;
+        info!(
+            "DIRECT HTTP RETRIEVAL SUCCESS: Retrieved {} bytes",
+            ciphertext.len()
+        );
+
+        // Verify digest
+        let calculated_digest = compute_digest(&ciphertext);
+        if calculated_digest != ciphertext_digest {
+            let calculated_digest_hex = hex::encode(&calculated_digest);
+            return Err(anyhow!(
+                "DIGEST MISMATCH: Expected: {digest_hex}, Got: {calculated_digest_hex}",
+            ));
```

Comment:
Is this a recoverable error, or will it cause the app to crash? If Yes - are you sure it's a desired behavior that cannot be exploited?